From fa9fb8f03b1942aa57742e413544aa0a981f7c8e Mon Sep 17 00:00:00 2001
From: Vara Prasad A V S G <vavsg@codeaurora.org>
Date: Tue, 12 Jun 2018 16:51:26 +0530
Subject: [PATCH 09/14] VNDK: Added required libs

Change-Id: I445c8a096b91d79a629beeca12234ca0d3aaa200
---
 power-660.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/power-660.c b/power-660.c
index e7daa08..d704705 100644
--- a/power-660.c
+++ b/power-660.c
@@ -38,6 +38,8 @@
 #include <fcntl.h>
 #include <dlfcn.h>
 #include <stdlib.h>
+#include <pthread.h>
+#include <unistd.h>
 
 #define LOG_TAG "QCOM PowerHAL"
 #include <log/log.h>
-- 
2.17.1

