From 8a123617ceba74ce70eed2d70dd2dabf66415ed3 Mon Sep 17 00:00:00 2001
From: Vara Prasad A V S G <vavsg@codeaurora.org>
Date: Tue, 12 Jun 2018 16:51:26 +0530
Subject: [PATCH 10/18] VNDK: Added required libs

Change-Id: I445c8a096b91d79a629beeca12234ca0d3aaa200
---
 power-660.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/power-660.c b/power-660.c
index ac1137e..f9f60e1 100644
--- a/power-660.c
+++ b/power-660.c
@@ -37,6 +37,8 @@
 #include <fcntl.h>
 #include <dlfcn.h>
 #include <stdlib.h>
+#include <pthread.h>
+#include <unistd.h>
 
 #define LOG_TAG "QCOM PowerHAL"
 #include <log/log.h>
-- 
2.17.1

