From 54b6830a97869495ae6e1e9b7d4cbc2e9da6d1c7 Mon Sep 17 00:00:00 2001
From: Bharat Pawar <bpawar@codeaurora.org>
Date: Tue, 26 Jun 2018 13:06:17 +0530
Subject: [PATCH 12/16] VNDK: Added required headers for 8998 target

Change-Id: I7e2143b4c29ba9f0e4fc8ddbe1d700a2b17d513a
---
 power-8998.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/power-8998.c b/power-8998.c
index af14e67..28b7c38 100644
--- a/power-8998.c
+++ b/power-8998.c
@@ -38,6 +38,7 @@
 #include <dlfcn.h>
 #include <stdlib.h>
 #include <pthread.h>
+#include <unistd.h>
 
 #define LOG_TAG "QCOM PowerHAL"
 #include <log/log.h>
-- 
2.17.1

