From 9af2699b857fbc1b8e44e3b88eb3859a8c3efd76 Mon Sep 17 00:00:00 2001
From: nx111 <gd.zhangdz@gmail.com>
Date: Wed, 6 Dec 2017 08:33:10 +0800
Subject: check version in project directory.

Change-Id: Ie0165545274f6db14e1c803a073becb72382ba8b

diff --git a/version.sh b/version.sh
index 3b7cafd..def966d 100755
--- a/version.sh
+++ b/version.sh
@@ -3,7 +3,7 @@
 VERSION="4.1"
 OUT="$1"
 
-if [ -d .git ] && head=`git rev-parse --verify HEAD 2>/dev/null`; then
+if [ -d .git -a -f iw.c ] && head=`git rev-parse --verify HEAD 2>/dev/null`; then
 	git update-index --refresh --unmerged > /dev/null
 	descr=$(git describe --match=v*)
 
