From f68547b140ca9a684ff1c3745a3fe11220638e1b Mon Sep 17 00:00:00 2001
From: Michael Bestas <mkbestas@lineageos.org>
Date: Mon, 28 May 2018 13:40:43 +0300
Subject: [PATCH] thermal: Use project pathmap

Change-Id: I029e0cba36d7c8a730429363350f62a00208a8bb
---
 Android.mk | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Android.mk b/Android.mk
index 1a14500..5b2e190 100644
--- a/Android.mk
+++ b/Android.mk
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+ifeq ($(call my-dir),$(call project-path-for,qcom-thermal))
+
 LOCAL_PATH := $(call my-dir)
 
 include $(CLEAR_VARS)
@@ -37,3 +39,5 @@ LOCAL_MODULE_TAGS := optional
 LOCAL_CFLAGS := -Wno-unused-parameter
 
 include $(BUILD_SHARED_LIBRARY)
+
+endif
-- 
2.17.1

