From 8932104eaab2bfec8cfd490e6a834884f90411bc Mon Sep 17 00:00:00 2001
From: codeworkx <daniel.hillenbrand@codeworkx.de>
Date: Wed, 22 Aug 2018 14:24:30 +0200
Subject: [PATCH 3/3] Snap: use platform cert

Change-Id: I23f73723a9b62ee67b6d0f234898828eba945962
---
 Android.mk | 1 +
 1 file changed, 1 insertion(+)

diff --git a/Android.mk b/Android.mk
index 0af8c52c1..572fafeec 100644
--- a/Android.mk
+++ b/Android.mk
@@ -35,6 +35,7 @@ LOCAL_JAVA_LIBRARIES := org.lineageos.platform.internal
 
 LOCAL_PACKAGE_NAME := Snap
 LOCAL_PRIVILEGED_MODULE := true
+LOCAL_CERTIFICATE := platform
 
 LOCAL_AAPT_FLAGS += --rename-manifest-package org.lineageos.snap
 
-- 
2.17.1

