From f3ec426d97467d551bd6027544079ed151e538c2 Mon Sep 17 00:00:00 2001
From: Tom Marshall <tdm.code@gmail.com>
Date: Fri, 29 Mar 2019 21:17:15 +0000
Subject: [PATCH 3/3] recovery: Set SELinux to permissive for non-release
 builds

Change-Id: Iad020439a007a28351719774b641038b076601a2
---
 recovery.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/recovery.cpp b/recovery.cpp
index b9598583..2eead17b 100644
--- a/recovery.cpp
+++ b/recovery.cpp
@@ -1679,6 +1679,11 @@ int main(int argc, char **argv) {
     return 1;
   }
 
+#ifndef RELEASE_BUILD
+  // Set SELinux to permissive
+  security_setenforce(0);
+#endif
+
   // Clear umask for packages that copy files out to /tmp and then over
   // to /system without properly setting all permissions (eg. gapps).
   umask(0);
-- 
2.17.1

