From fd5d006367217a3beb35326d5ce8c9d84c2c95a0 Mon Sep 17 00:00:00 2001
From: Tom Marshall <tdm.code@gmail.com>
Date: Fri, 29 Mar 2019 21:17:15 +0000
Subject: [PATCH 4/5] recovery: Set SELinux status to Permissive for recovery
 images

Change-Id: Iad020439a007a28351719774b641038b076601a2
---
 recovery.cpp | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/recovery.cpp b/recovery.cpp
index b9598583..c7f47700 100644
--- a/recovery.cpp
+++ b/recovery.cpp
@@ -1641,6 +1641,9 @@ int main(int argc, char **argv) {
   // log to stdout (which is redirected to recovery.log) as we used to do.
   android::base::InitLogging(argv, &UiLogger);
 
+  // Set SELinux to Permissive status
+  security_setenforce(0);
+
   // Take last pmsg contents and rewrite it to the current pmsg session.
   static const char filter[] = "recovery/";
   // Do we need to rotate?
-- 
2.17.1

