From 965a37fa01b71801d20a144ec99a62d1d6cb5dfe Mon Sep 17 00:00:00 2001
From: nx111 <gd.zhangdz@gmail.com>
Date: Sat, 9 Dec 2017 21:11:47 +0800
Subject: networkMonitor: add fallback captive urls for china.

Change-Id: I53ddd959c52f29bf6c664bb4b66196f69d431b38

diff --git a/services/core/java/com/android/server/connectivity/NetworkMonitor.java b/services/core/java/com/android/server/connectivity/NetworkMonitor.java
index 11296ab..f3acd96 100644
--- a/services/core/java/com/android/server/connectivity/NetworkMonitor.java
+++ b/services/core/java/com/android/server/connectivity/NetworkMonitor.java
@@ -92,7 +92,7 @@ public class NetworkMonitor extends StateMachine {
             "http://connectivitycheck.gstatic.com/generate_204";
     private static final String DEFAULT_FALLBACK_URL  = "http://www.google.com/gen_204";
     private static final String DEFAULT_OTHER_FALLBACK_URLS =
-            "http://play.googleapis.com/generate_204";
+            "https://www.google.cn/generate_204,https://www.qualcomm.cn/generate_204,https://www.v2ex.com/generate_204,http://play.googleapis.com/generate_204";
     private static final String DEFAULT_USER_AGENT    = "Mozilla/5.0 (X11; Linux x86_64) "
                                                       + "AppleWebKit/537.36 (KHTML, like Gecko) "
                                                       + "Chrome/52.0.2743.82 Safari/537.36";
