From 70090fed197c9a5366c35981af409f1f87e04072 Mon Sep 17 00:00:00 2001
From: nx111 <gd.zhangdz@gmail.com>
Date: Sun, 26 Nov 2017 17:12:57 +0800
Subject: move mixe_paths.xml to /system/etc/

Change-Id: Ia00b16ce85c8f27db090378f0974db7cc07de179

diff --git a/hal/msm8974/platform.c b/hal/msm8974/platform.c
index 6a79960..1e29bc5 100644
--- a/hal/msm8974/platform.c
+++ b/hal/msm8974/platform.c
@@ -40,9 +40,9 @@
 
 #define SOUND_TRIGGER_DEVICE_HANDSET_MONO_LOW_POWER_ACDB_ID (100)
 
-#define MIXER_XML_PATH "/vendor/etc/mixer_paths.xml"
-#define MIXER_XML_PATH_AUXPCM "/vendor/etc/mixer_paths_auxpcm.xml"
-#define MIXER_XML_PATH_WCD9330 "/vendor/etc/mixer_paths_wcd9330.xml"
+#define MIXER_XML_PATH "/system/etc/mixer_paths.xml"
+#define MIXER_XML_PATH_AUXPCM "/system/etc/mixer_paths_auxpcm.xml"
+#define MIXER_XML_PATH_WCD9330 "/system/etc/mixer_paths_wcd9330.xml"
 #define LIB_ACDB_LOADER "libacdbloader.so"
 #define AUDIO_DATA_BLOCK_MIXER_CTL "HDMI EDID"
 
