From 5fae0897ae4eb0180b28b564670e62446b360022 Mon Sep 17 00:00:00 2001
From: nx111 <gd.zhangdz@gmail.com>
Date: Mon, 27 Nov 2017 00:51:46 +0800
Subject: move platform info xml to /system/etc

Change-Id: Id3a1dd28772ebc6cd4bdeccb1d0bdb03e586a610

diff --git a/hal/platform_info.c b/hal/platform_info.c
index 6edc175..48f19ab 100644
--- a/hal/platform_info.c
+++ b/hal/platform_info.c
@@ -38,7 +38,7 @@
 #include "platform_api.h"
 #include <platform.h>
 
-#define PLATFORM_INFO_XML_PATH      "/vendor/etc/audio_platform_info.xml"
+#define PLATFORM_INFO_XML_PATH      "/system/etc/audio_platform_info.xml"
 #define BUF_SIZE                    1024
 
 typedef enum {
