From 02a7be2d87b5e03453da96cd7a78ea28efc21dfa Mon Sep 17 00:00:00 2001
From: nx111 <gd.zhangdz@gmail.com>
Date: Fri, 21 Sep 2018 11:38:39 +0800
Subject: [PATCH 7/7] soong: generated_kernel_headers defined in vendor/lineage

Change-Id: I07c270c6415f10904b78847c48549c67713cf89d
---
 Android.bp | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/Android.bp b/Android.bp
index b9350210..64ed4db1 100644
--- a/Android.bp
+++ b/Android.bp
@@ -360,12 +360,6 @@ kernel_headers {
     vendor: true,
 }
 
-generated_headers {
-    name: "generated_kernel_headers",
-    additional_deps: ["$(TARGET_OUT_INTERMEDIATES)/KERNEL_OBJ/usr"],
-    vendor: true,
-}
-
 cc_genrule {
     name: "host_bionic_linker_asm",
     host_supported: true,
-- 
2.17.1

