From 60c66ba0376352ad4b8450b97f7ecce010199459 Mon Sep 17 00:00:00 2001
From: nx111 <gd.zhangdz@gmail.com>
Date: Wed, 3 Oct 2018 21:33:30 +0800
Subject: [PATCH 4/4] klte-common: use legacy pthread mutex init to fix RIL.

Change-Id: I6aa978571d4f36928328e0e7681eff9219e1c51c
---
 BoardConfigCommon.mk | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/BoardConfigCommon.mk b/BoardConfigCommon.mk
index 5bdbf7c..8bde9af 100644
--- a/BoardConfigCommon.mk
+++ b/BoardConfigCommon.mk
@@ -23,6 +23,9 @@ TARGET_SPECIFIC_HEADER_PATH := $(COMMON_PATH)/include
 # Audio
 USE_CUSTOM_AUDIO_POLICY := 1
 
+# Bionic
+TARGET_USES_LEGACY_PTHREAD_MUTEX := true
+
 # Bluetooth
 BOARD_BLUETOOTH_BDROID_BUILDCFG_INCLUDE_DIR := $(COMMON_PATH)/bluetooth
 BOARD_CUSTOM_BT_CONFIG := $(COMMON_PATH)/bluetooth/vnd_klte.txt
-- 
2.17.1

