From 0e24de68ed3b50a6f8b52e884d7404774b05f10d Mon Sep 17 00:00:00 2001
From: Rashed Abdel-Tawab <rashed@linux.com>
Date: Sun, 12 Aug 2018 19:08:06 -0700
Subject: [PATCH] nano: don't spam warnings as errors

Change-Id: I83a6e3d926ebd3241a92a00ac4d3926b1999c931
---
 Android.mk | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/Android.mk b/Android.mk
index 1327ff7b..014d5d75 100644
--- a/Android.mk
+++ b/Android.mk
@@ -33,6 +33,8 @@ LOCAL_CFLAGS += \
 	-DLOCALEDIR=\"/data/locale\" \
 	-DSYSCONFDIR=\"/system/etc/nano\"
 
+LOCAL_CFLAGS += -Wno-sign-compare -Wno-unused-parameter
+
 LOCAL_SHARED_LIBRARIES += \
 	libncurses \
 	libssh
-- 
2.17.1

