From c989e27a60a59c6f8333325ee57a7420ea3264e7 Mon Sep 17 00:00:00 2001
From: Harsh Shandilya <msfjarvis@gmail.com>
Date: Sat, 17 Dec 2016 23:29:25 +0530
Subject: [PATCH] envsetup: stop jack server once build completed

Change-Id: I5b53890b65f31a0d098a1ddd353a5151bb52b437
---
 envsetup.sh | 1 +
 1 file changed, 1 insertion(+)

diff --git a/envsetup.sh b/envsetup.sh
index 2987b13..91e065d 100644
--- a/envsetup.sh
+++ b/envsetup.sh
@@ -1678,6 +1678,7 @@ function _wrap_build()
     fi
     echo " ####${color_reset}"
     echo
+    $(gettop)/prebuilts/sdk/tools/jack-admin stop-server 2>&1 >/dev/null
     return $ret
 }
 
-- 
2.7.4

