From a7f84279ecf2fc1f2a84abaa95802708668bbaa2 Mon Sep 17 00:00:00 2001
From: Harsh Shandilya <msfjarvis@gmail.com>
Date: Sat, 17 Dec 2016 23:29:25 +0530
Subject: [PATCH 8/8] envsetup: stop jack server once build completed

Change-Id: I5b53890b65f31a0d098a1ddd353a5151bb52b437
---
 envsetup.sh | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/envsetup.sh b/envsetup.sh
index 2987b13..40639d2 100644
--- a/envsetup.sh
+++ b/envsetup.sh
@@ -31,6 +31,7 @@ Environment options:
 - SANITIZE_HOST: Set to 'true' to use ASAN for all host modules. Note that
                  ASAN_OPTIONS=detect_leaks=0 will be set by default until the
                  build is leak-check clean.
+- ANDROID_HOST_JACK_DESTROY: Set to 'true' to stop the Jack server after builds.
 
 Look at the source to view more functions. The complete list is:
 EOF
@@ -1678,6 +1679,9 @@ function _wrap_build()
     fi
     echo " ####${color_reset}"
     echo
+    if [ ! -z "$ANDROID_HOST_JACK_DESTROY" ]; then
+        $(gettop)/prebuilts/sdk/tools/jack-admin stop-server 2>&1 >/dev/null
+    fi
     return $ret
 }
 
-- 
2.7.4

