From a495739351dd27102b6c64b1972a915945350568 Mon Sep 17 00:00:00 2001
From: Barry Clark <barry@barryclark.co>
Date: Sat, 21 Jun 2014 09:33:25 -0400
Subject: [PATCH 013/214] Update default github icon to point to jekyll now
 repo

---
 _config.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/_config.yml b/_config.yml
index fbec663..7888d29 100644
--- a/_config.yml
+++ b/_config.yml
@@ -21,7 +21,7 @@ footer-links:
   email: 
   facebook: 
   flickr: 
-  github: jekyll
+  github: barryclark/jekyll-now
   instagram: 
   linkedin: 
   pinterest: 
-- 
2.17.1

