From c2d379cac670b45c4769f2126db79825c0570467 Mon Sep 17 00:00:00 2001
From: Barry Clark <barry@barryclark.co>
Date: Wed, 30 Jul 2014 19:09:58 -0400
Subject: [PATCH 030/214] Clean out grunt stuff from _config.yml excludes

---
 _config.yml | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/_config.yml b/_config.yml
index 75a0f91..5f99ab5 100644
--- a/_config.yml
+++ b/_config.yml
@@ -69,5 +69,5 @@ gems:
 exclude:
     - Gemfile
     - Gemfile.lock
-    - scss
+    - LICENSE
     - README.md
\ No newline at end of file
-- 
2.17.1

