From 615f8ca5f3af9fd31312d0e01215f9fa6e189ab2 Mon Sep 17 00:00:00 2001
From: Barry Clark <barry@barryclark.co>
Date: Fri, 15 Aug 2014 15:31:47 -0400
Subject: [PATCH 048/214] Add permalink with trailing slash to about page Fixes
 #13

---
 about.md | 1 +
 1 file changed, 1 insertion(+)

diff --git a/about.md b/about.md
index 22b8692..db54e4d 100644
--- a/about.md
+++ b/about.md
@@ -1,6 +1,7 @@
 ---
 layout: page
 title: About
+permalink: /about/
 ---
 
 Some information about you!
-- 
2.17.1

