From 282a3c0b586b5463e54cf8db765b92f62ed591da Mon Sep 17 00:00:00 2001
From: Barry Clark <barry@barryclark.co>
Date: Mon, 23 Feb 2015 01:06:50 -0500
Subject: [PATCH 075/214] License date bump

---
 LICENSE | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/LICENSE b/LICENSE
index 282d036..40e6030 100644
--- a/LICENSE
+++ b/LICENSE
@@ -1,6 +1,6 @@
 The MIT License (MIT)
 
-Copyright (c) 2014 Barry Clark
+Copyright (c) 2015 Barry Clark
 
 Permission is hereby granted, free of charge, to any person obtaining a copy of
 this software and associated documentation files (the "Software"), to deal in
-- 
2.17.1

