From d58e0a965c81de06422e9a53a87a49266272e9ea Mon Sep 17 00:00:00 2001
From: Tom Powell <zifnab@zifnab06.net>
Date: Wed, 21 Dec 2016 15:56:02 -0800
Subject: [PATCH 114/214] Fix header image

---
 _config.yml           | 2 --
 _layouts/default.html | 2 +-
 2 files changed, 1 insertion(+), 3 deletions(-)

diff --git a/_config.yml b/_config.yml
index fabbd9d..7229f0b 100644
--- a/_config.yml
+++ b/_config.yml
@@ -8,8 +8,6 @@ name: Lineage OS
 # Short bio or description (displayed in the header)
 description: Lineage OS Android Distribution
 
-# URL of your avatar or profile pic (you could use your GitHub profile pic)
-avatar: /images/android.png
 
 #
 # Flags below are optional
diff --git a/_layouts/default.html b/_layouts/default.html
index 00976e5..b405357 100644
--- a/_layouts/default.html
+++ b/_layouts/default.html
@@ -19,7 +19,7 @@
     <div class="wrapper-masthead">
       <div class="container">
         <header class="masthead clearfix">
-          <a href="{{ site.baseurl }}/" class="site-avatar"><img src="{{ site.avatar }}" /></a>
+          <a href="{{ site.baseurl }}/" class="site-avatar"><img src="{{site.baseurl}}/images/android.png" /></a>
 
           <div class="site-info">
             <h1 class="site-name"><a href="{{ site.baseurl }}/">{{ site.name }}</a></h1>
-- 
2.17.1

