From 381d89599374fac6c04529ef5aa6682bbc90f088 Mon Sep 17 00:00:00 2001
From: Tom Powell <zifnab@zifnab06.net>
Date: Sat, 24 Dec 2016 14:55:26 -0800
Subject: [PATCH 117/214] Update CNAME

---
 CNAME | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/CNAME b/CNAME
index 8b13789..edc7838 100644
--- a/CNAME
+++ b/CNAME
@@ -1 +1 @@
-
+lineageos.org
\ No newline at end of file
-- 
2.17.1

