From 29fcefa31a89c98b6824e2c3224d00c1be13ef87 Mon Sep 17 00:00:00 2001
From: Tom Powell <zifnab@zifnab06.net>
Date: Sun, 25 Dec 2016 22:19:35 -0800
Subject: [PATCH 120/214] Add status page

---
 _layouts/default.html | 1 +
 style.scss            | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/_layouts/default.html b/_layouts/default.html
index b405357..62c7a70 100644
--- a/_layouts/default.html
+++ b/_layouts/default.html
@@ -30,6 +30,7 @@
             <a href="{{ site.baseurl }}/">Blog</a>
             <a href="{{ site.baseurl }}/about">About</a>
             <a href="{{ site.baseurl }}/community">Community</a>
+            <a href="http://status.lineageos.org">Status</a>
           </nav>
         </header>
       </div>
diff --git a/style.scss b/style.scss
index 6f46710..2842cf7 100644
--- a/style.scss
+++ b/style.scss
@@ -214,7 +214,7 @@ nav {
   float: right;
   margin-top: 23px; // @TODO: Vertically middle align
   font-family: $helveticaNeue;
-  font-size: 18px;
+  font-size: 16px;
 
   @include mobile {
     float: none;
-- 
2.17.1

