From 7c7a26b22b819b56f4dc5f4ce8722e31a3cd3803 Mon Sep 17 00:00:00 2001
From: Stricted <info@stricted.net>
Date: Mon, 24 Sep 2018 20:50:26 +0200
Subject: [PATCH 15/15] dtbhtool: fix unused variable 'model' warning

Change-Id: I0ac347428d01789622487d803ae13d433240063c
---
 dtbhtool/dtbimg.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dtbhtool/dtbimg.c b/dtbhtool/dtbimg.c
index cddcd29..f751e88 100644
--- a/dtbhtool/dtbimg.c
+++ b/dtbhtool/dtbimg.c
@@ -126,7 +126,9 @@ void *load_dtbh_block(const char *dtb_path, unsigned pagesize, unsigned *_sz)
     uint32_t version = DTBH_VERSION;
     unsigned blob_sz = 0;
     char fname[PATH_MAX];
+#ifdef DTBH_MODEL
     const unsigned *model;
+#endif
     const unsigned *prop_chip;
     const unsigned *prop_platform;
     const unsigned *prop_subtype;
-- 
2.17.1

