From b49c268149e77e3a1b1db1aba71568f814479b03 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Nikola=20Majki=C4=87?= <n1kolaa924@gmail.com>
Date: Thu, 5 Jan 2017 06:12:54 +0000
Subject: [PATCH 074/249] Remove whitespace

Change-Id: I5987a56ad93d0b65a0200508833e38596a8e9f6d
---
 README.mkdn | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/README.mkdn b/README.mkdn
index 23b6ebd..37742ae 100644
--- a/README.mkdn
+++ b/README.mkdn
@@ -9,7 +9,7 @@ You can do this by using these commands:
     (From root android directory)
     . build/envsetup.sh
     (Go to repo you are patching, make your changes and commit)
-    cmgerrit <for(new)/changes(patch set)> <branch/change-id> 
+    cmgerrit <for(new)/changes(patch set)> <branch/change-id>
 
     repo start cm-14.1 .
     (Make your changes and commit)
@@ -39,7 +39,7 @@ Then to sync up:
 
 Please see the [LineageOS Wiki](http://wiki.lineageos.org/) for building instructions.
 
-For more information on this Github Organization and how it is structured, 
+For more information on this Github Organization and how it is structured,
 please [read the wiki article](http://wiki.lineageos.org/w/Github_Organization)
 
 Buildbot
-- 
2.17.1

