From aff6d7288025af058dca3021b3a9fedadcf2d608 Mon Sep 17 00:00:00 2001
From: Luca Stefani <luca.stefani.ge1@gmail.com>
Date: Sun, 26 Aug 2018 19:18:14 +0200
Subject: [PATCH 11/14] su: Remove useless casts

Change-Id: I037c90c39e161392d2a9d344075c8c5d9da033f7
---
 daemon.c | 2 +-
 su.c     | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/daemon.c b/daemon.c
index 6686c7e..80f735b 100644
--- a/daemon.c
+++ b/daemon.c
@@ -155,7 +155,7 @@ static void send_fd(int sockfd, int fd) {
 static int read_int(int fd) {
     int val;
     int len = read(fd, &val, sizeof(int));
-    if (len != (int)sizeof(int)) {
+    if (len != sizeof(int)) {
         ALOGE("unable to read int: %d", len);
         exit(-1);
     }
diff --git a/su.c b/su.c
index b0cfed9..808be9e 100644
--- a/su.c
+++ b/su.c
@@ -65,7 +65,7 @@ static int from_init(struct su_initiator* from) {
     }
 
     /* Get the command line */
-    snprintf(path, sizeof(path), "/proc/%u/cmdline", (unsigned)from->pid);
+    snprintf(path, sizeof(path), "/proc/%d/cmdline", from->pid);
     fd = open(path, O_RDONLY);
     if (fd < 0) {
         PLOGE("Opening command line");
@@ -102,7 +102,7 @@ static int from_init(struct su_initiator* from) {
     }
 
     /* If this isn't app_process, use the real path instead of argv[0] */
-    snprintf(path, sizeof(path), "/proc/%u/exe", (unsigned)from->pid);
+    snprintf(path, sizeof(path), "/proc/%d/exe", from->pid);
     len = readlink(path, exe, sizeof(exe));
     if (len < 0) {
         PLOGE("Getting exe path");
-- 
2.17.1

