From f3cdb7be84f5d902d21c71c9533e73338057d743 Mon Sep 17 00:00:00 2001
From: topjohnwu <topjohnwu@gmail.com>
Date: Mon, 18 Dec 2017 15:46:01 +0800
Subject: [PATCH 12/12] su: Initialize windows size

Change-Id: I00de6d71b377e4c34387a7e8f571d985e0ebc011
---
 pts.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/pts.c b/pts.c
index 12a47c3..dd94c7b 100644
--- a/pts.c
+++ b/pts.c
@@ -218,10 +218,8 @@ static void* watch_sigwinch(void* data) {
     sigemptyset(&winch);
     sigaddset(&winch, SIGWINCH);
 
+    // Wait for a SIGWINCH
     do {
-        // Wait for a SIGWINCH
-        sigwait(&winch, &sig);
-
         if (closing_time) break;
 
         // Get the new terminal size
@@ -233,7 +231,7 @@ static void* watch_sigwinch(void* data) {
         // Set the new terminal size
         ioctl(slave, TIOCSWINSZ, &w);
 
-    } while (1);
+    } while (sigwait(&winch, &sig) == 0);
 
     free(data);
     return NULL;
@@ -288,8 +286,6 @@ int watch_sigwinch_async(int master, int slave) {
         return -1;
     }
 
-    // Set the initial terminal size
-    raise(SIGWINCH);
     return 0;
 }
 
-- 
2.17.1

